### ISSUE TITLE
Automatically infer class variables based on declaration

### ISSUE THREAD

**Issue by FeldrinH** [microsoft/pylance-release#4389]
There seems to be a limitation with Pylance detecting if a variable declared in a class is a class variable or not.

In particular, the following code:
```python
from typing import ClassVar, Protocol

class SimpleProtocol(Protocol):
    a: ClassVar[str]

class SimpleClass:
    a = "test"

val: SimpleProtocol = SimpleClass()
```

Gives the following error:
![image](https://github.com/microsoft/pylance-release/assets/5256211/fb266c17-ed0d-48b0-81d3-34d34196a1f5)
As far as I understand, `SimpleClass.a` is in fact a class variable (and using it as such does not produce any runtime errors),

Changing `SimpleClass` definition to
```python
class SimpleClass:
    a: ClassVar = "test"
```
gets rid of the error.

It would be nice if Pylance could automatically infer that `a` is a class variable without the need for explicit type annotations.

Tested with Pylance language server 2023.5.30 (pyright e1b6074b).

**Comment by FeldrinH** [comment]
I'm a little confused.

If I understand you correctly, then this declares `SimpleProtocol.a` to be a regular class variable:
```
class SimpleProtocol(Protocol):
    a: str
```

Yet, with this declaration the following gives an error:
```python
from typing import Protocol, Type

class SimpleProtocol(Protocol):
    a: str

def black_box() -> Type[SimpleProtocol]:
    raise NotImplementedError

typ: Type[SimpleProtocol] = black_box()
print(typ.a)
```
![image](https://github.com/microsoft/pylance-release/assets/5256211/381a6b5e-4aec-4734-8d4b-2cdfa83305be)

And the following does not give an error:
```python
from typing import Protocol

class SimpleProtocol(Protocol):
    a: str

class SimpleClass:
    def __init__(self):
        self.a = "test"

val: SimpleProtocol = SimpleClass()
```

So in what sense is `SimpleProtocol.a` a 'regular class variable'? In both these examples the type checking seems to behave as if `a` was a pure instance variable.

**Comment by erictraut** [comment]
[PEP 544](https://peps.python.org/pep-0544/) (which introduced Protocol classes) doesn't provide a way to distinguish between the three different types of class-level variables. It introduced `ClassVar`, which is clearly meant for "pure class variables". For other variables, it seems to indicate that they should be treated as pure instance variables. 

> To distinguish between protocol class variables and protocol instance variables, the special ClassVar annotation should be used as specified by [PEP 526](https://peps.python.org/pep-0526)

However, now that I re-read that statement, I realize that it leaves some room for interpretation.

I just checked with mypy, another type checker that was the reference implementation for PEP 544, and it appears to use the following interpretation for protocols:

```python
class SimpleProtocol(Protocol):
    a: ClassVar[str] # Pure class variable
    b: str # Regular class variable

    def __init__(self):
        self.c: str # Pure instance variable
```

Currently, pyright treats `b` as a pure instance variable within a protocol (consistent with my reading of the PEP), and that's what's leading to the error you're seeing.

It would be more consistent if protocol and non-protocol classes worked in a consistent manner, and if that doesn't violate PEP 544 (as it apparently does not), then I'm happy to make that change in pyright. That would also make pyright consistent with mypy in this regard.

Could someone on the pylance team please transfer this issue to the pyright repo?

**Comment by FeldrinH** [comment]
FWIW the example in my original post also typechecks in Pycharm with no errors and no warnings.

**Comment by FeldrinH** [comment]
Also probably worth noting:
The mypy interpretation of class and instance variables seems to introduce a different issue.

The following code passes `mypy --strict` with no issues found, but raises an error at runtime when accessing `typ.a`:
```python
from typing import Protocol, Type

class SimpleProtocol(Protocol):
    a: str

class SimpleClass:
    def __init__(self) -> None:
        self.a = "test"

def black_box() -> Type[SimpleProtocol]:
    return SimpleClass

typ: Type[SimpleProtocol] = black_box()
print(typ.a)
```

**Comment by erictraut** [comment]
Yes, that's a general problem with "regular class variables" in Python. There's generally no way for a static type checker to know whether they have been initialized at the class or instance level (or both). Most other programming languages (for good reason) force you to declare whether a variable is a class or instance variable, and they don't permit it to be both.

**Comment by FeldrinH** [comment]
Fair enough. The main question that I am still unclear on is this:

If Pyright was changed to consider the following code as valid and well typed, would it violate some PEP or lead to problematic soundness issues?
```python
from typing import ClassVar, Protocol

class SimpleProtocol(Protocol):
    a: ClassVar[str]

class SimpleClass:
    a = "test"

val: SimpleProtocol = SimpleClass()
```

**Comment by erictraut** [comment]
> If Pyright was changed to consider the following code as valid and well typed, would it violate some PEP or lead to problematic soundness issues?

I think this once again depends on how PEP 544 is interpreted. Looking to mypy (which, as I mentioned, was the reference implementation for PEP 544), it generates an error in this case. I'm struggling to reconcile mypy's behaviors. It seems to treat `a = "test"` as a "pure instance variable" in this case (which doesn't make sense because it's clearly being assigned as a class variable). It interprets it as a "regular class variable" in other cases.

Setting aside mypy's inconsistent behavior and thinking about this from first principles, I agree with your assertion that this should be sound from a typing perspective. `SimpleClass.a` should be interpreted as a "regular class variable", which satisfies the "pure class variable" defined in the protocol. (Note: the converse situation would not be type safe.)

My conclusion is that there are two changes that I should make in pyright:
1. Remove the special-case handling for variables declared within the class body of a protocol class so they are categorized as "regular class variables" (as opposed to "pure instance variables") if they do not include a `ClassVar` annotation.
2. When performing protocol type consistency checks, allow a "regular class variable" to satisfy a protocol that defines the symbol as a "pure class variable".

Does that sound right to you?

**Comment by FeldrinH** [comment]
That does sound right to me. For the first change I don't really have a strong opinion either way. I fully agree with the second change and would be very happy if it was implemented.

**Comment by erictraut** [comment]
I'm going to hold off on the first change and give it more thought.

I've implemented the second change, and this will be in next week's release of pyright (and the insider build of pylance).

**Comment by gramster** [comment]
Did you decide about the first change Eric? Wondering if we can close this now.

**Comment by erictraut** [comment]
Apologies, I forgot about this issue. After thinking about this more, I think we should go ahead and make the first change as well. This will be included in the next release of pyright — and the next insiders build of pylance. You can close this issue after this week's pylance is released.

**Comment by gramster** [comment]
Thanks!

**Comment by heejaechang** [comment]
This issue has been fixed in prerelease version 2023.6.31, which we've just released. You can find the changelog here: [CHANGELOG.md](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#2023631-21-June-2023-PreRelease)
